src/charm.py

```diff
@@ -359,14 +372,25 @@ def _are_db_relations_ready(self) -> bool:
         Returns:
             If the needed relations have been established.
         """
-        # mypy fails do detect this stored value can be False
-        if not self._stored.db_name:  # type: ignore
+        db_rel = self._database.get_relation_data()
+        if db_rel is None:
             self.model.unit.status = WaitingStatus("Waiting for database relation")
             return False
+        if None in (
+            db_rel["POSTGRES_USER"],
+            db_rel["POSTGRES_PASSWORD"],
+            db_rel["POSTGRES_DB"],
+            db_rel["POSTGRES_HOST"],
```

Comment:
`POSTGRES_HOST` is never going to be `None`, btw, since you default to `":".split(":")[0]`.

Reply:
Should I leave it like this, as defensive code (against eventual changes in src/database.py) ?

Reply:
I'd prefer to see this check handled in `src/database.py` directly, but if we're leaving it here for now, I think we should update this to say if the other values are `None` or if this one is `''`.

Reply:
I've moved most of the related logic to the database module (and improved the consistency of the get_relation_data() function).